src/System.Numerics.Tensors/System/Numerics/Tensor.cs

```diff
@@ -0,0 +1,1197 @@
+        /// Initializes a rank-n Tensor using the dimensions specified in <paramref name="dimensions"/>.
+        /// </summary>
+        /// <param name="dimensions"></param>
+        public Tensor(params int[] dimensions)
+        {
+            if (dimensions == null)
+            {
+                throw new ArgumentNullException(nameof(dimensions));
+            }
+
+            if (dimensions.Length == 0)
+            {
+                // rank 0 tensor should be a scalar, need to think about how to represent that.
+                throw new ArgumentOutOfRangeException(nameof(dimensions));
+            }
+
+            long size = GetProduct(dimensions);
+
+            // could throw, let the runtime decide what that limit is
+            backingArray = new T[size];
+
+            // make a private copy of mutable dimensions array
+            this.dimensions = new int[dimensions.Length];
+            dimensions.CopyTo(this.dimensions, 0);
+        }
+
+        public Tensor(T[] fromBackingArray, params int[] dimensions)
+        {
+            if (dimensions == null)
+            {
+                throw new ArgumentNullException(nameof(dimensions));
+            }
+
+            if (dimensions.Length == 0)
+            {
+                // rank 0 tensor should be a scalar, need to think about how to represent that.
+                throw new ArgumentOutOfRangeException(nameof(dimensions));
+            }
+
+            long size = GetProduct(dimensions);
+
+            if (size != fromBackingArray.Length)
+            {
+                throw new ArgumentException($"Length of {nameof(fromBackingArray)} ({fromBackingArray.Length}) must match product of {nameof(dimensions)} ({size}).");
+            }
+
+            // keep a reference to the backing array
+            backingArray = fromBackingArray;
+
+            // make a private copy of mutable dimensions array
+            this.dimensions = new int[dimensions.Length];
+            dimensions.CopyTo(this.dimensions, 0);
+        }
+
+        /// <summary>
+        /// Total length of the Tensor.
+        /// </summary>
+        public int Length => backingArray.Length;
+
+        /// <summary>
+        /// Rank of the tensor: number of dimensions.
+        /// </summary>
+        public int Rank => dimensions.Length;
+
+        /// <summary>
+        /// Returns a copy of the dimensions array.
+        /// </summary>
+        public IReadOnlyList<int> Dimensions => readOnlyDimensions ?? (readOnlyDimensions = new ReadOnlyCollection<int>(dimensions));
+
+        /// <summary>
+        /// Returns a single dimensional view of this Tensor, in C-style ordering
+        /// </summary>
+        public T[] Buffer => backingArray;
+
+        /// <summary>
+        /// Sets all elements in Tensor to <paramref name="value"/>.
+        /// </summary>
+        /// <param name="value">Value to fill</param>
+        public void Fill(T value)
+        {
+            // JIT look here, lend us a hand
```

Comment:
According to a post by the JIT team a while back, they recognize this pattern explicitly and generate better code for it.